### DISCUSSION TITLE
关于性能测试数据与对比的若干问题

### DISCUSSION THREAD

**Discussion by aryu1986** [bytedance/monoio#123] [Q&A]
你好，首先这个异步框架太酷了，值得持续关注。但性能对比上有几点疑问想咨询下：

1.Tokio本身作为通用的异步运行时，肯定有work-steal的过程；目前这个性能对比中这部分还是对其有影响；我想问下是不是有可能做一个带 work-steal的Monoio做一些对比呢？

2.整个对比图中没有内存消耗的对比数据，这部分的情况如何，在同等任务和压力的情况下，我们的内存开销和Tokio比是否也有优势？

3.假设撇开work-steal和thread-per-core，以及epoll和io_uring的差异，我们和Tokio在设计上还有哪些不同的优化使得我们的性能更优？是否方便介绍下

**Comment by ihciah** [comment]
1. task 是否在固定的线程运行会影响非常多的设计。一是 runtime 内部对 task 的约束、和很多代码实现都要改，所以支持这个约等于重写了。二是实际上这个行为也是影响用户代码，所以很难对比实际影响。比如，在我们的 runtime 中你可以很方便地使用 thread-local，但如果在有 work stealing 的 runtime 下，你的 task 会被 schedule 到不同的 thread 上，所以很难有效利用 thread-local。比如要做数据聚合功能（例如限流器、打点上报等），可以直接按线程聚合，每个线程上定期收集；但在 work stealing 下就不得不引入锁，每个 task 去抢锁然后修改。
2. 内存没有对比，但理论上应该没太大差别。
3. 主要的差异就在你说的这两点了，实际上 runtime 本身是一个很轻的实现，可以优化的地方没那么多。实际上我们很多设计和代码都是借鉴的 tokio，我认为 tokio 的设计是很优雅的，只是和我们目标场景不同。其他的 diff 可能更倾向于我们独有的功能性的东西，比如跨线程唤醒等（任务可以等待其他线程上的唤醒事件，但还是在原有线程执行）。
